File: JANUARY/08-01-2025/Count Subarray with k odds.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @returns {number}
 */
class Solution {
    countSubarrays(arr, k) {
        const countMap = new Map();
        countMap.set(0, 1);
        let prefixSum = 0;
        let result = 0;
        for (let num of arr) {
            if (num % 2 !== 0) prefixSum += 1;
            if (countMap.has(prefixSum - k)) {
                result += countMap.get(prefixSum - k);
            }
            countMap.set(prefixSum, (countMap.get(prefixSum) || 0) + 1);
        }
        return result;
    }
}

File: JANUARY/08-01-2025/Count Subarray with k odds.java
import java.util.HashMap;

class Solution {
    public int countSubarrays(int[] arr, int k) {
        HashMap<Integer, Integer> countMap = new HashMap<>();
        countMap.put(0, 1);
        int prefixSum = 0;
        int result = 0;
        for (int num : arr) {
            if (num % 2 != 0) {
                prefixSum += 1;
            }
            result += countMap.getOrDefault(prefixSum - k, 0);
            countMap.put(prefixSum, countMap.getOrDefault(prefixSum, 0) + 1);
        }
        return result;
    }
}